debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/TransactionCommitConsumer.java

```diff
@@ -109,233 +102,383 @@ public void accept(LogMinerEvent event) throws InterruptedException {
-            return event.getScn().equals(prevEvent.getScn()) && event.getRsId().equals(prevEvent.getRsId());
+            offset += length - newLength;
+            length = newLength;
         }
-        return false;
     }
 
-    private boolean isSameTableRow(LogMinerEvent event, LogMinerEvent prevEvent) {
-        final Table table = schema.tableFor(event.getTableId());
-        if (table == null) {
-            LOGGER.trace("Unable to locate table '{}' schema, unable to detect if same row.", event.getTableId());
-            return false;
+    private static class LobUnderConstruction {
+        final List<LobFragment> fragments = new LinkedList<>();
+        int start = 0;
+        int end = 0;
+        boolean binary = false;
+        boolean isNull = true; // result of #merge() should be null (for instances that are never written to)
+
+        void add(LobFragment fragment) {
+            isNull = false;
+
+            if (fragments.isEmpty()) { // first fragment to be added
+                fragments.add(fragment);
+                start = fragment.offset;
+                end = fragment.end();
+                binary = fragment.binary;
+                return;
+            }
+
+            if (fragment.binary != binary) {
+                throw new DebeziumException("mixing binary and non-binary writes in a single LOB");
+            }
+
+            if (fragment.offset >= end) { // the expected case
+                fragments.add(fragment);
+                end = fragment.end();
+                return;
+            }
+
+            // the uncommon case: writing somewhere in the middle
+
+            // find the right spot to insert
+            ListIterator<LobFragment> iter = fragments.listIterator();
+            while (iter.hasNext()) {
+                LobFragment frag = iter.next();
+                if (fragment.offset < frag.end() && fragment.offset >= frag.offset) {
+                    // truncate frag and insert after
+                    frag.truncate(fragment.offset - frag.offset);
+                    iter.add(fragment);
+                    break;
+                }
+                if (frag.offset > fragment.offset) {
+                    // insert before; no need to truncate preceding fragment
+                    iter.previous();
+                    iter.add(fragment);
+                    break;
+                }
+            }
+
+            // are there any following fragments that are (partially) overwritten by the fragment we're inserting?
+            while (iter.hasNext()) {
+                LobFragment frag = iter.next();
+                if (frag.offset >= fragment.end()) { // we're done
+                    break;
+                }
+                if (frag.end() <= fragment.end()) { // remove entirely
+                    iter.remove();
+                }
+                else { // front-truncate
+                    frag.frontTruncate(frag.end() - fragment.end());
+                }
+            }
```

Comment:
Have you measured the performance impact of this iteration at all?  Particularly when the size of the file/character data grows to a respectable size of 5MB, 10MB, 25MB?  Are there any optimizations that could be done here to avoid or minimize the iteration per fragment added?  I'm just concerned this could introduce performance problems when each fragment is writing 8k or 16k of data in a large 10MB or 25MB file.

Reply:
It's the uncommon case (the expected case is all-appends, which is handled efficiently above), so I didn't spend too much time thinking about optimizing this, in the interest of keeping the code simple and understandable.
If you want, I can look into this.

Actually, I now see that I missed a corner case in this algorithm: when a new fragment overwrites the middle of an existing fragment (say, there is a fragment with offset 1000, length 1000 and we add a new fragment with offset 1400, length 100), the remainder of the original fragment (offsets 1500-2000) will be truncated erroneously. I'll go back to the drawing board to fix that, and perhaps add an additional test for this case.

Reply:
Sounds great.